### PR TITLE
support upper triangular matrices in linalg

### PR DESCRIPTION
**Pull request by asmushetzel**
## Description ##
Allows use of upper triangular matrices in the linalg operators potrf/potri/trmm/trsm. Usually all of these operators allow specification of whether we deal with lower or upper triangular matrices in BLAS/LAPACK. We haven't exposed this so far in MXNet and only supported lower triangular ones. But it turns out that there are indeed applications where you would like to use upper triangular ones directly. 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X ] Changes are complete (i.e. I finished coding on this PR)
- [ X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by asmushetzel** [comment]
@mseeger

**Comment by Roshrini** [comment]
@asmushetzel Thanks for your contribution.
@mxnet-label-bot [pr-awaiting-review]
@apeforest @samskalicky Requesting your review.

File: src/operator/tensor/la_op_inline.h [hunk 1]
```diff
@@ -35,17 +35,23 @@ using namespace mshadow;
 // Helper functions.
 struct CopyLowerToUpper {
```
[the remark this answers is not in the document]
Comment (mseeger) [reply src/operator/tensor/la_op_inline.h]: OK, I see below that using this inverse argument is quite elegant.
But consider renaming the function to something like CopyTriangleToOppositeTriangle,
and then inverse -> source_upper, or so?

File: src/operator/tensor/la_op_inline.h [hunk 2]
```diff
@@ -35,17 +35,23 @@ using namespace mshadow;
...
+    if (row > col) {
+       if (inverse) {
+         data[i] = data[i + (col - row) * (stride - 1)];
+       } else {
+         data[i + (col - row) * (stride - 1)] = data[i];
+       }
+    }
   }
 };
 struct ZeroUpper {
```
[the remark this answers is not in the document]
Comment (mseeger) [reply src/operator/tensor/la_op_inline.h]: Maybe rename to ZeroTriangle, and then inverse -> lower?

File: src/operator/tensor/la_op_inline.h [hunk 3]
```diff
@@ -403,15 +413,18 @@ struct potrf_backward {
...
     // by setting the upper triangle to be equal to the lower triangle, leaving
     // lower triangle and diagonal unchanged.
+    // The function also handles the case when L is upper triangular by appropriate
+    // transpositions.
+    const LaCholeskyParam& param = nnvm::get<LaCholeskyParam>(attrs.parsed);
     if ( dL.dptr_ != dA.dptr_ ) {
       Copy(dA, dL, s);
     }
-    trmm::op(L, dA, DType(1.0), false, true, s);
+    trmm::op(L, dA, DType(1.0), !param.lower, param.lower, true, s);
```
Comment (mseeger) [inline src/operator/tensor/la_op_inline.h]: This looks good to me. Very elegant!

File: src/operator/tensor/la_op_inline.h [hunk 4]
```diff
@@ -430,14 +443,22 @@ struct potri_backward {
...
     // dL = -tril( A * (dA + dA**T) * L**(-T)), where tril() extracts lower triangle
     // and diagonal. We must not assume that dA is symmetric.
+    // The function also handles the case when L is upper triangular by appropriate
+    // transpositions.
     // Note: Calling gemm twice here is a bit wasteful, but otherwise the symmetrization
     // of dA would require temporary memory.
-    gemm::op(A, dA, dL, DType(1.), DType(0.), false, false, s);
-    gemm::op(A, dA, dL, DType(1.), DType(1.), false, true, s);
-    trsm::op(L, dL, DType(-1.), true, true, s);
+    const LaCholeskyParam& param = nnvm::get<LaCholeskyParam>(attrs.parsed);
```
Comment (mseeger) [inline src/operator/tensor/la_op_inline.h]: Looks good to me

File: src/operator/tensor/la_op_inline.h [hunk 5]
```diff
@@ -458,14 +479,15 @@ struct trsm_backward {
     const LaTriangMatrixMultParam& param = nnvm::get<LaTriangMatrixMultParam>(attrs.parsed);
```
Comment (mseeger) [inline src/operator/tensor/la_op_inline.h]: Looks good to me

File: src/operator/tensor/la_op_inline.h [hunk 6]
```diff
@@ -494,10 +516,10 @@ struct trmm_backward {
     }
     using namespace mxnet_op;
     Kernel<ZeroUpper, xpu>::Launch(s, dL.MSize(), dL.size(1)*dL.stride_, dL.stride_,
-                                   dL.dptr_);
+                                   dL.dptr_, !param.lower);
```
Comment (mseeger) [inline src/operator/tensor/la_op_inline.h]: Looks good to me

File: tests/python/unittest/test_operator.py [hunk 7]
```diff
@@ -5034,7 +5034,7 @@ def _make_symm_symbol(a, ndims):
     tr_shape = tuple(tr_shape)
     return 0.5 * (a + mx.sym.transpose(a, axes=tr_shape))
 
-def _make_lower_triangle_symm(a, ndims, m, dtype=np.float32):
+def _make_lower_triangle_symm(a, ndims, m, lower, dtype=np.float32):
```
Comment (mseeger) [inline tests/python/unittest/test_operator.py]: Could change name to make_triangle_symm ?

File: tests/python/unittest/test_operator.py [hunk 8]
```diff
@@ -5186,141 +5189,147 @@ def test_laop():
         check_grad(test_gemm, [a2, b2])
 
     # Now test all the other operators.
+    for lower in [True, False]:
+        upper = not lower
```
Comment (mseeger) [inline tests/python/unittest/test_operator.py]: Do you need this?

**Comment by mseeger** [comment]
@asmushetzel 
One other point: Could you mention our ArXiV paper in the code, referring to it for all the detailed derivations?
https://arxiv.org/abs/1710.08717

**Comment by asmushetzel** [comment]
@mseeger Addressed all your suggestions. Thanks a lot

**Comment by asmushetzel** [comment]
Travis timed out when running tests. Unrelated to this change. Rest looks ok now.

**Comment by asmushetzel** [comment]
The reference to the underlying paper is already in la_op_inline.h as a \note in the header section.

**Comment by ankkhedia** [comment]
@sandeep-krishnamurthy This PR looks good. Could you please add pr-waiting-merge and merge it.

**Comment by sandeep-krishnamurthy** [comment]
@samskalicky - FYI. Can you please take a look?

File: src/operator/tensor/la_op.cc [hunk 9]
```diff
@@ -227,10 +231,11 @@ NNVM_REGISTER_OP(_linalg_potri)
...
-If *n=2*, *A* is a lower triangular matrix (entries of upper triangle are all zero)
+If *n=2*, *A* is a triangular matrix (entries of upper or lower triangle are all zero)
 with positive diagonal. We compute:
 
-  *out* = *A*\ :sup:`-T` \* *A*\ :sup:`-1`
+  *out* = *A*\ :sup:`-T` \* *A*\ :sup:`-1` if *lower* = *true*
+  *out* = *A*\ :sup:`-1` \* *A*\ :sup:`-T` if *lower* = *false*
 
 In other words, if *A* is the Cholesky factor of a symmetric positive definite matrix
 *B* (obtained by *potrf*), then
```
Comment (samskalicky) [inline src/operator/tensor/la_op.cc]: This is confusing, the example is *out* = *A*\ :sup:`-T` \* *A*\ :sup:`-1` so "out" is function of "A", and then here you refer to "B", whats "B"? Do you mean "out"? Can you use the same format as above:
*A* = *B* \* *B*\ :sup:`T`
 where A is the output and B is the input for consistency?

File: src/operator/tensor/la_op.h [hunk 10]
```diff
@@ -81,10 +81,21 @@ struct LaMatrixMultParam : public dmlc::Parameter<LaMatrixMultParam> {
   }
 };
 
+// Parameters for Cholesky factorization and matrix inversion
+struct LaCholeskyParam : public dmlc::Parameter<LaCholeskyParam> {
+  bool lower;
+  DMLC_DECLARE_PARAMETER(LaCholeskyParam) {
+    DMLC_DECLARE_FIELD(lower)
+      .set_default(true)
+      .describe("Triangular matrix is lower triangular.");
```
Comment (samskalicky) [inline src/operator/tensor/la_op.h]: Would it be clearer in the description to say: "'lower' is true if matrix is lower triangular or false for upper triangular" to remove ambiguity?

File: src/operator/tensor/la_op.h [hunk 11]
```diff
@@ -93,6 +104,9 @@ struct LaTriangMatrixMultParam : public dmlc::Parameter<LaTriangMatrixMultParam>
     DMLC_DECLARE_FIELD(rightside)
       .set_default(false)
       .describe("Multiply triangular matrix from the right to non-triangular one.");
+    DMLC_DECLARE_FIELD(lower)
+      .set_default(true)
+      .describe("Triangular matrix is lower triangular.");
```
Comment (samskalicky) [inline src/operator/tensor/la_op.h]: same as above

File: src/operator/tensor/la_op_inline.h [hunk 12]
```diff
@@ -403,15 +417,18 @@ struct potrf_backward {
     // Here, copyLTU(M) creates a symmetric matrix from the square matrix M
     // by setting the upper triangle to be equal to the lower triangle, leaving
     // lower triangle and diagonal unchanged.
+    // The function also handles the case when L is upper triangular by appropriate
+    // transpositions.
```
Comment (samskalicky) [inline src/operator/tensor/la_op_inline.h]: Can you change the variable names here to be consistent with the new doc? change dL and L to dB and B both in the comment and in the op(args) for the function.

File: src/operator/tensor/la_op_inline.h [hunk 13]
```diff
@@ -430,14 +447,22 @@ struct potri_backward {
     // Backward of A = potri(L).
     // dL = -tril( A * (dA + dA**T) * L**(-T)), where tril() extracts lower triangle
     // and diagonal. We must not assume that dA is symmetric.
+    // The function also handles the case when L is upper triangular by appropriate
+    // transpositions.
```
Comment (samskalicky) [inline src/operator/tensor/la_op_inline.h]: same as above L ==> B

File: src/operator/tensor/la_op_inline.h [hunk 14]
```diff
@@ -458,14 +483,15 @@ struct trsm_backward {
     const LaTriangMatrixMultParam& param = nnvm::get<LaTriangMatrixMultParam>(attrs.parsed);
     // Compute dA
     if ( dA.dptr_ != dB.dptr_ ) Copy(dA, dB, s);
-    trsm::op(L, dA, DType(param.alpha), param.rightside, !param.transpose, s);
+    trsm::op(L, dA, DType(param.alpha), param.rightside, param.lower, !param.transpose, s);
```
Comment (samskalicky) [inline src/operator/tensor/la_op_inline.h]: same as above L==>B

File: src/operator/tensor/la_op_inline.h [hunk 15]
```diff
@@ -493,11 +519,11 @@ struct trmm_backward {
       gemm::op(A, dB, dL, scale, DType(0.), !param.transpose, param.transpose, s);
     }
     using namespace mxnet_op;
-    Kernel<ZeroUpper, xpu>::Launch(s, dL.MSize(), dL.size(1)*dL.stride_, dL.stride_,
-                                   dL.dptr_);
+    Kernel<ZeroTriangular, xpu>::Launch(s, dL.MSize(), dL.size(1)*dL.stride_, dL.stride_,
```
Comment (samskalicky) [inline src/operator/tensor/la_op_inline.h]: same as above, rename L

**Comment by samskalicky** [comment]
added some suggestions on cleaning up the names throughout to match the new documentation. This is necessary for maintenance and future debugging. The jenkins tests are passing, so it looks good to me.

File: src/operator/tensor/la_op.cc (hunk 9 above)
Comment (asmushetzel) [reply src/operator/tensor/la_op.cc]: The name of the input of the function is currently "A" in FListInputNames() so changing this would make the operator interface non-backward compatible. So we have to stick with it and then use also "A" accordingly in the documentation as the input to this function. Note that it is explained below that the output "out" is the inverse of a matrix "B" if the input "A" is the Cholesky-factor of "B". 
I tried a few rewordings, but none of them made this clearer.

File: src/operator/tensor/la_op.h (hunk 10 above)
Comment (asmushetzel) [reply src/operator/tensor/la_op.h]: Good suggestion. Done.

File: src/operator/tensor/la_op.h (hunk 11 above)
Comment (asmushetzel) [reply src/operator/tensor/la_op.h]: dito

File: src/operator/tensor/la_op_inline.h (hunk 12 above)
Comment (asmushetzel) [reply src/operator/tensor/la_op_inline.h]: Will change L/dL for potrf/potri to B/dB.  Though this will trigger a bigger renaming as "L" for lower triangular is also used in trsm/trmm and this would be then inconsistent too. So for the latter functions, I did a remapping of B-->C, A-->B, L-->A which actually brings this internal naming in line with the external documentation (having "C" denoting "out").

File: src/operator/tensor/la_op.cc (hunk 9 above)
Comment (samskalicky) [reply src/operator/tensor/la_op.cc]: thanks for the explanation, makes sense.

**Comment by asmushetzel** [comment]
Please note that someone else just renamed la_op_inline.h today to la_op-inl. So all conversation related to that file lost pointers.

**Comment by asmushetzel** [comment]
@samskalicky Incorporated all (possible) suggestions for variable renaming.

**Comment by ankkhedia** [comment]
@asmushetzel Could you please look into CI failure. At first glance, it looks like failure in Scala package. if the error seems to be unrelated, could you please re-trigger the CI by forcing an empty commit.

**Comment by asmushetzel** [comment]
Travis errors out because of time limit again.
Jenkins fails because of some recipe that the scala GPU tests did not find.
Anyone who can fix these issues that seem to be related to build system only?

**Comment by ankkhedia** [comment]
@asmushetzel Could you plese retrigger the CI by an empty commit. The error might be flaky and may not come up again if you re-trigger the build.

**Comment by ankkhedia** [comment]
It seems like this PR will solve the issue 
https://github.com/apache/incubator-mxnet/pull/13081

**Comment by asmushetzel** [comment]
Indeed Jenkins runs through now. Travis times out as usual. 
No further actions planned from my side.

**Comment by ankkhedia** [comment]
@sandeep-krishnamurthy @nswamy Could you please take a look and merge this PR if it looks good.

**Comment by lupesko** [comment]
@anirudh2290 can you please help with a review/merge? thanks!

**Comment by ankkhedia** [comment]
> @asmushetzel seems that the OSX Travis build is failing. Can you look into it?

@lupesko Travis build is not a blocker for PR merge currently.

**Comment by asmushetzel** [comment]
Is there anything left that is expected from my side?
